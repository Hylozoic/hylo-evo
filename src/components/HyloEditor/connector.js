import { connect } from 'react-redux'
import { findMentions, clearMentions, getMentionResults } from './store'

export function mapStateToProps (state, props) {
  return {
    mentionResults: getMentionResults(state, props)
  }
}

export const mapDispatchToProps = {
  findMentions,
  clearMentions
}

export default connect(mapStateToProps, mapDispatchToProps)
