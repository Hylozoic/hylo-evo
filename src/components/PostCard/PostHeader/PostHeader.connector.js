import { connect } from 'react-redux'
import { push } from 'react-router-redux'
import { removePostFromUrl, editPostUrl } from 'util/navigation'
import getMe from 'store/selectors/getMe'
import {
  deletePost,
  removePost,
  pinPost,
  getCommunity
} from './PostHeader.store'

export function mapStateToProps (state, props) {
  const community = getCommunity(state, props)

  return {
    currentUser: getMe(state, props),
    community
  }
}

export function mapDispatchToProps (dispatch, props) {
  const { slug } = props.routeParams
  const closeUrl = removePostFromUrl(window.location.pathname)
  const deletePostWithConfirm = postId => {
    if (window.confirm('Are you sure you want to delete this post?')) {
      dispatch(deletePost(postId))
      .then(() => dispatch(push(closeUrl)))
    }
  }

  return {
    editPost: postId => props.editPost
      ? props.editPost(postId)
      : dispatch(push(editPostUrl(postId, props.routeParams))),
    deletePost: postId => props.deletePost
      ? props.deletePost(postId)
      : deletePostWithConfirm(postId),
    removePost: postId => props.removePost
      ? props.removePost(postId)
      : dispatch(removePost(postId, slug)),
    pinPost: (postId, communityId) => props.pinPost
      ? props.pinPost(postId)
      : dispatch(pinPost(postId, communityId))
  }
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  const { currentUser, community } = stateProps
  const { id, creator } = ownProps
  const { deletePost, editPost, removePost, pinPost } = dispatchProps
  const isCreator = currentUser && creator && currentUser.id === creator.id
  const canEdit = isCreator
  const canModerate = currentUser && currentUser.canModerate(community)

  return {
    ...stateProps,
    ...dispatchProps,
    ...ownProps,
    deletePost: isCreator ? () => deletePost(id) : undefined,
    editPost: canEdit ? () => editPost(id) : undefined,
    canFlag: !isCreator,
    pinPost: canModerate && community ? () => pinPost(id, community.id) : undefined,
    removePost: !isCreator && canModerate ? () => removePost(id) : undefined,
    canEdit
  }
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)
