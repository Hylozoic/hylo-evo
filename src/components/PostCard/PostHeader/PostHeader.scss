.header {
  padding: 0px 15px;
  margin-bottom: 30px;
}

.headerMainRow {
  display: flex;
  justify-content: flex-start;
  align-items: center;
}

.authorName {
  /* !empty */
}

.avatar {
  margin-right: 10px;
}

.headerText {
  display: inline-block;
  position: relative;
  top: 1px;
  line-height: 15px;
}

.userName {
  // TODO: Extract typography?
  @include font-bold;
  color: $color-rhino-80;
  font-size: 14px;
  &:hover {
    color: $color-rhino;
    text-decoration: none;
  }
}

.userTitle {
  // TODO: Extract typography? this one is very close to bdy-drk-sm
  @include font-regular;
  color: $color-rhino-80;
  font-size: 14px;
}

.date {
  line-height: 12px;
}

.spacer {
  composes: timestamp from 'css/typography.scss';
  margin: 0 7px;
  font-size: 8px;
  position: relative;
  top: -2px;
}

.timestampRow {
  align-items: center;
  display: flex;
}

.timestamp {
  composes: timestamp from 'css/typography.scss';
}

.context {
  // TODO: Extract typography?
  @include font-regular;
  color: $color-caribbean-green;
  font-size: 12px;
  &:hover {
    color: $color-gossamer;
    text-decoration: none;
  }
}

.upperRight {
  margin-left: auto;
}

.pinIcon {
  top: 1px;
  margin-right: 12px;
  font-size: 24px;
}

.label {
  margin-right: 20px;
  // TODO remove this positioning. Only necessary because of Dropdown breaking layout
  position: relative;
  top: -3px;
}

.menuLink:hover {
  text-decoration: none;
}

.menuLink:hover > span {
  color: $color-rhino-80;
}

.close {
  display: inline-block;
  cursor: pointer;
  position: relative;
  top: 2px;
}

.topicsLine {
  composes: text-button from 'css/typography.scss';
  font-size: 12px;
  display: inline-block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  max-width: 220px;
}

.topic {
  composes: text-button from 'css/typography.scss';
  font-size: 12px;
  margin-right: 10px;
}

.newLineForTopics {
  margin-left: 10px;
  margin-top: 10px;
  width: 450px;
}
