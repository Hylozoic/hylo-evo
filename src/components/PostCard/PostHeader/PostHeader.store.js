import orm from 'store/models'
import { createSelector as ormCreateSelector } from 'redux-orm'

export const MODULE_NAME = 'PostHeader'

// Constants
export const DELETE_POST = `${MODULE_NAME}/DELETE_POST`
export const DELETE_POST_PENDING = DELETE_POST + '_PENDING'
export const REMOVE_POST = `${MODULE_NAME}/REMOVE_POST`
export const REMOVE_POST_PENDING = REMOVE_POST + '_PENDING'
export const PIN_POST = `${MODULE_NAME}/PIN_POST`
export const PIN_POST_PENDING = `${PIN_POST}_PENDING`

// Action Creators
export function deletePost (id) {
  return {
    type: DELETE_POST,
    graphql: {
      query: `mutation ($id: ID) {
        deletePost(id: $id) {
          success
        }
      }`,
      variables: {
        id
      }
    },
    meta: {
      optimistic: true,
      id
    }
  }
}

export function removePost (postId, slug) {
  return {
    type: REMOVE_POST,
    graphql: {
      query: `mutation ($postId: ID, $slug: String) {
        removePost(postId: $postId, slug: $slug) {
          success
        }
      }`,
      variables: {
        postId,
        slug
      }
    },
    meta: {
      optimistic: true,
      postId,
      slug
    }
  }
}

export function pinPost (postId, communityId) {
  return {
    type: PIN_POST,
    graphql: {
      query: `mutation ($postId: ID, $communityId: ID) {
        pinPost(postId: $postId, communityId: $communityId) {
          success
        }
      }`,
      variables: {
        postId,
        communityId
      }
    },
    meta: {
      optimistic: true,
      postId,
      communityId
    }
  }
}

export const getCommunity = ormCreateSelector(
  orm,
  state => state.orm,
  (_, { slug }) => slug,
  (session, slug) => session.Community.safeGet({slug})
)

export function ormSessionReducer ({ Post }, { type, meta }) {
  var post
  switch (type) {
    case DELETE_POST_PENDING:
      Post.withId(meta.id).delete()
      break

    case REMOVE_POST_PENDING:
      post = Post.withId(meta.postId)
      const communities = post.communities.filter(c =>
        c.slug !== meta.slug).toModelArray()
      post.update({communities})
      break

    case PIN_POST_PENDING:
      post = Post.withId(meta.postId)
      // this line is to clear the selector memoization
      post.update({_invalidate: (post._invalidate || 0) + 1})
      let postMembership = post.postMemberships.filter(p =>
        Number(p.community) === Number(meta.communityId)).toModelArray()[0]
      postMembership && postMembership.update({pinned: !postMembership.pinned})
  }
}
