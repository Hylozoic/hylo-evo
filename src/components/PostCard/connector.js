import { connect } from 'react-redux'
import { push } from 'react-router-redux'

export function mapStateToProps (state, props) {
  return {}
}

export const mapDispatchToProps = {
  navigate: push
}

export default connect(mapStateToProps, mapDispatchToProps)
