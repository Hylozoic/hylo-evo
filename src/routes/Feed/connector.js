import { connect } from 'react-redux'
import { SAMPLE_FEED_ITEMS } from './sampleData'

function mapStateToProps (state) {
  return {
    feedItems: SAMPLE_FEED_ITEMS
  }
}

function mapDispatchToProps (dispatch, props) {
  return {}
}

export default connect(mapStateToProps, mapDispatchToProps)
