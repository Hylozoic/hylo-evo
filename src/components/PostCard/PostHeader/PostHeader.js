import React from 'react'
import { Link } from 'react-router-dom'
import Avatar from 'components/Avatar'
import Dropdown from 'components/Dropdown'
import PostLabel from 'components/PostLabel'
import Highlight from 'components/Highlight'
import Icon from 'components/Icon'
import { communityUrl, personUrl } from 'util/index'
import { humanDate } from 'hylo-utils/text'
import './PostHeader.scss'
import { filter, isFunction } from 'lodash'

export default function PostHeader ({
  creator,
  date,
  type,
  pinned,
  communities,
  close,
  className,
  slug,
  showCommunity,
  editPost,
  deletePost,
  removePost,
  pinPost,
  highlightProps
}) {
  if (!creator) return null

  let context

  if (showCommunity) {
    const { name, slug } = communities[0]
    context = {
      label: name,
      url: communityUrl(slug)
    }
  }

  const dropdownItems = filter([
    // Leaving these here as they will be implemented in the future
    // {icon: 'Flag', label: 'Flag', onClick: () => console.log('Flag')},
    {icon: 'Pin', label: pinned ? 'Unpin' : 'Pin', onClick: pinPost},
    {icon: 'Edit', label: 'Edit', onClick: editPost},
    {icon: 'Trash', label: 'Delete', onClick: deletePost},
    {icon: 'Trash', label: 'Remove From Community', onClick: removePost}
    // {icon: 'Complete', label: 'Accept and mark complete', onClick: () => console.log('Accept and mark complete')}
  ], item => isFunction(item.onClick))

  return <div styleName='header' className={className}>
    <Avatar avatarUrl={creator.avatarUrl} url={personUrl(creator.id, slug)} styleName='avatar' />
    <div styleName='headerText'>
      <Highlight {...highlightProps}>
        <Link to={personUrl(creator.id, slug)} styleName='userName'>{creator.name}{creator.tagline && ', '}</Link>
      </Highlight>
      {creator.tagline && <span styleName='userTitle'>{creator.tagline}</span>}
      <div>
        <span styleName='timestamp'>
          {humanDate(date)}{context && <span styleName='spacer'>â€¢</span>}
        </span>
        {context && <Link to={context.url} styleName='context'>
          {context.label}
        </Link>}
      </div>
    </div>
    <div styleName='upperRight'>
      {pinned && <Icon name='Pin' styleName='pinIcon' />}
      {type && <PostLabel type={type} styleName='label' />}
      {dropdownItems.length > 0 &&
        <Dropdown toggleChildren={<Icon name='More' />} items={dropdownItems} />}
      {close && <a styleName='close' onClick={close}><Icon name='Ex' /></a>}
    </div>
  </div>
}
