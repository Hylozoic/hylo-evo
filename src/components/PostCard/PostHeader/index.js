import React from 'react'
import { Link } from 'react-router-dom'
import Avatar from 'components/Avatar'
import Dropdown from 'components/Dropdown'
import PostLabel from 'components/PostLabel'
import Icon from 'components/Icon'
import { communityUrl, personUrl } from 'util/index'
import { humanDate } from 'hylo-utils/text'
import './PostHeader.scss'

export default function PostHeader ({
  creator,
  date,
  type,
  communities,
  close,
  className,
  slug,
  showCommunity
}) {
  let context

  if (showCommunity) {
    const { name, slug } = communities[0]
    context = {
      label: name,
      url: communityUrl(slug)
    }
  }

  return <div styleName='header' className={className}>
    <Avatar avatarUrl={creator.avatarUrl} url={personUrl(creator.id, slug)} styleName='avatar' />
    <div styleName='headerText'>
      <Link to={personUrl(creator.id, slug)} styleName='userName'>{creator.name}{creator.tagline && ', '}</Link>
      {creator.tagline && <span styleName='userTitle'>{creator.tagline}</span>}
      <div>
        <span styleName='timestamp'>
          {humanDate(date)}{context && <span styleName='spacer'>â€¢</span>}
        </span>
        {context && <Link to={context.url} styleName='context'>
          {context.label}
        </Link>}
      </div>
    </div>
    <div styleName='upperRight'>
      {type && <PostLabel type={type} styleName='label' />}
      <Dropdown toggleChildren={<Icon name='More' />} items={[
        {icon: 'Pin', label: 'Pin', onClick: () => console.log('Pin')},
        {icon: 'Flag', label: 'Flag', onClick: () => console.log('Flag')},
        {icon: 'Trash', label: 'Delete', onClick: () => console.log('Delete')},
        {label: 'Other'},
        {icon: 'Complete', label: 'Accept and mark complete', onClick: () => console.log('Accept and mark complete')}
      ]} />
      {close && <a styleName='close' onClick={close}><Icon name='Ex' /></a>}
    </div>
  </div>
}
