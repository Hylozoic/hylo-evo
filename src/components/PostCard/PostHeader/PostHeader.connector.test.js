import { mapStateToProps, mapDispatchToProps, mergeProps } from './PostHeader.connector'
import orm from 'store/models'

describe('mapStateToProps', () => {
  it('maps', () => {
    const session = orm.session(orm.getEmptyState())
    session.Community.create({id: 33, slug: 'mycommunity'})
    session.Me.create({id: 20,
      memberships: [session.Membership.create({
        id: '345',
        community: 33,
        hasModeratorRole: true
      })]})

    const state = {
      orm: session.state
    }

    const ownProps = {creator: {id: 20}, routeParams: { slug: 'mycommunity' }}
    const { community, currentUser } = mapStateToProps(state, ownProps)

    expect(community.id).toBe(33)
    expect(currentUser.id).toBe(20)
  })
})

describe('mapDispatchToProps', () => {
  it('maps the action generators', () => {
    window.confirm = jest.fn()
    const dispatch = jest.fn(val => val)
    const props = {
      routeParams: {
        slug: 'mycommunity'
      }
    }
    const dispatchProps = mapDispatchToProps(dispatch, props)
    expect(dispatchProps).toMatchSnapshot()
    dispatchProps.removePost(10)
    dispatchProps.editPost(10)
    expect(dispatch).toHaveBeenCalledTimes(2)

    dispatchProps.deletePost(1)
    dispatchProps.pinPost(2, 3)
    expect(window.confirm).toHaveBeenCalledWith('Are you sure you want to delete this post?')
    expect(dispatch.mock.calls).toMatchSnapshot()
  })
})

describe('mergeProps', () => {
  const dispatchProps = {
    removePost: jest.fn(),
    deletePost: jest.fn(),
    editPost: jest.fn(),
    pinPost: jest.fn()
  }

  beforeEach(() => {
    dispatchProps.removePost.mockReset()
    dispatchProps.editPost.mockReset()
    dispatchProps.deletePost.mockReset()
    dispatchProps.pinPost.mockReset()
  })

  describe('as moderator', () => {
    it('can delete and edit own posts, and pin a post', () => {
      const session = orm.session(orm.getEmptyState())
      const community = session.Community.create({id: 33, slug: 'mycommunity'})
      session.Me.create({id: 20,
        memberships: [session.Membership.create({
          id: '345',
          community: community.id,
          hasModeratorRole: true
        })]})
      const state = {
        orm: session.state
      }
      const ownProps = {id: 20, routeParams: { slug: 'mycommunity' }, creator: {id: 20}}
      const stateProps = mapStateToProps(state, ownProps)
      const { deletePost, removePost, editPost, canEdit, pinPost } = mergeProps(stateProps, dispatchProps, ownProps)

      expect(canEdit).toBeTruthy()
      expect(deletePost).toBeTruthy()
      expect(removePost).toBeFalsy()
      expect(editPost).toBeTruthy()

      deletePost()
      expect(dispatchProps.deletePost).toHaveBeenCalledWith(20)

      editPost()
      expect(dispatchProps.editPost).toHaveBeenCalledWith(20)

      pinPost()
      expect(dispatchProps.pinPost).toHaveBeenCalledWith(20, 33)
    })

    it('cannot delete posts but can moderate', () => {
      const session = orm.session(orm.getEmptyState())
      const community = session.Community.create({id: 33, slug: 'mycommunity'})
      session.Me.create({id: 20,
        memberships: [session.Membership.create({
          id: '345',
          community: community.id,
          hasModeratorRole: true
        })]})

      const state = {
        orm: session.state
      }

      const ownProps = {id: 20, routeParams: { slug: 'mycommunity' }, creator: {id: 33}}
      const stateProps = mapStateToProps(state, ownProps)

      const {deletePost, removePost, editPost} = mergeProps(stateProps, dispatchProps, ownProps)

      expect(deletePost).toBeFalsy()
      expect(editPost).toBeFalsy()
      expect(removePost).toBeTruthy()

      removePost()
      expect(dispatchProps.removePost).toHaveBeenCalledWith(20)
    })
  })

  describe('as non-moderator', () => {
    it("can delete own posts, can't pin posts", () => {
      const session = orm.session(orm.getEmptyState())
      const community = session.Community.create({id: 33, slug: 'mycommunity'})
      session.Me.create({id: 20,
        memberships: [session.Membership.create({
          id: '345',
          community: community.id,
          hasModeratorRole: false
        })]})

      const state = {
        orm: session.state
      }

      const ownProps = {id: 20, routeParams: { slug: 'mycommunity' }, creator: {id: 20}}
      const stateProps = mapStateToProps(state, ownProps)

      const {deletePost, removePost, editPost, pinPost} = mergeProps(stateProps, dispatchProps, ownProps)

      expect(editPost).toBeTruthy()
      expect(deletePost).toBeTruthy()
      expect(removePost).toBeFalsy()
      expect(pinPost).toBeFalsy()

      deletePost()
      expect(dispatchProps.deletePost).toHaveBeenCalledWith(20)
    })
  })

  it('cannot delete or remove posts if not creator or moderator', () => {
    const session = orm.session(orm.getEmptyState())
    const community = session.Community.create({id: 33, slug: 'mycommunity'})
    session.Me.create({id: 20,
      memberships: [session.Membership.create({
        id: '345',
        community: community.id,
        hasModeratorRole: false
      })]})

    const state = {
      orm: session.state
    }

    const ownProps = {id: 20, routeParams: { slug: 'mycommunity' }, creator: {id: 33}}
    const stateProps = mapStateToProps(state, ownProps)

    const {deletePost, removePost, editPost} = mergeProps(stateProps, dispatchProps, ownProps)

    expect(deletePost).toBeFalsy()
    expect(removePost).toBeFalsy()
    expect(editPost).toBeFalsy()
  })
})
