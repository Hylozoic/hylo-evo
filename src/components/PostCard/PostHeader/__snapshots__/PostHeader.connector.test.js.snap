// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapDispatchToProps maps the action generators 1`] = `
Object {
  "deletePost": [Function],
  "editPost": [Function],
  "pinPost": [Function],
  "removePost": [Function],
}
`;

exports[`mapDispatchToProps maps the action generators 2`] = `
Array [
  Array [
    Object {
      "graphql": Object {
        "query": "mutation ($postId: ID, $slug: String) {
        removePost(postId: $postId, slug: $slug) {
          success
        }
      }",
        "variables": Object {
          "postId": 10,
          "slug": "mycommunity",
        },
      },
      "meta": Object {
        "optimistic": true,
        "postId": 10,
        "slug": "mycommunity",
      },
      "type": "PostHeader/REMOVE_POST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "args": Array [
          "/c/mycommunity/p/10/edit",
        ],
        "method": "push",
      },
      "type": "@@router/CALL_HISTORY_METHOD",
    },
  ],
  Array [
    Object {
      "graphql": Object {
        "query": "mutation ($postId: ID, $communityId: ID) {
        pinPost(postId: $postId, communityId: $communityId) {
          success
        }
      }",
        "variables": Object {
          "communityId": 3,
          "postId": 2,
        },
      },
      "meta": Object {
        "communityId": 3,
        "optimistic": true,
        "postId": 2,
      },
      "type": "PostHeader/PIN_POST",
    },
  ],
]
`;
